### PR TITLE
fix: correctly assign null if target is nullable instead of throwing

### LINKED ISSUE #1182: ThrowOnPropertyMappingNullMismatch throws an exception for nullable properties
**Describe the bug**
The XML comment of `ThrowOnPropertyMappingNullMismatch` says:
```xml
/// <summary>
/// Specifies the behaviour in the case when the mapper tries to set a non-nullable property to a <c>null</c> value.
/// If set to <c>true</c> an <see cref="ArgumentNullException"/> is thrown.
/// If set to <c>false</c> the property assignment is ignored.
/// This is ignored for required init properties and <see cref="IQueryable{T}"/> projection mappings.
/// </summary>
```

If I understood that right, if I have a non-nullable property in my `target` and the mapper tries to set it to `null`, that will result in an exception. However, that implies that if the target property is nullable, its value should be simply set to null. In my hypothetical scenario, I have the case where I have a nullable property in my target, but the mapper still throws an exception when it tries to set its value to `null`.

In the scenario below, the `UserDTO.Parent` property is a nullable (`UserDTO?`) property, but the generated `else` statement by the mapper throws an exception if my source (`User.Parent`) is `null`. From what it looks to me, that should not be the case, or did I misunderstand the `ThrowOnPropertyMappingNullMismatch` XML comment?

I am aware of `AllowNullPropertyAssignment` existence, but using it in a combination with `ThrowOnPropertyMappingNullMismatch` still throws the exception in this case.

**Declaration code**
```csharp
public class UserDTO
{
    public long? ParentId { get; set; }
    public UserDTO? Parent { get; set; }
}

public class User
{
    public long? ParentId { get; set; }
    public User? Parent { get; set; }
}

[Mapper(ThrowOnPropertyMappingNullMismatch = true)]
public partial class Mapper
{
    public partial UserDTO ToUserDTO(User user);
}
```

**Actual relevant generated code**
```csharp
[global::System.CodeDom.Compiler.GeneratedCode("Riok.Mapperly", "3.4.0.0")]
public partial global::UserDTO ToUserDTO(global::User user)
{
    var target = new global::UserDTO();
    if (user.Parent != null)
    {
        target.Parent = ToUserDTO(user.Parent);
    }
    else
    {
        throw new System.ArgumentNullException(nameof(user.Parent));
    }
    target.ParentId = user.ParentId;
    return target;
}
```

**Expected relevant generated code**
```csharp
[global::System.CodeDom.Compiler.GeneratedCode("Riok.Mapperly", "3.4.0.0")]
public partial global::UserDTO ToUserDTO(global::User user)
{
    var target = new global::UserDTO();
    if (user.Parent != null)
    {
        target.Parent = ToUserDTO(user.Parent);
    }
    else
    {
        target.Parent = null;
    }
    target.ParentId = user.ParentId;
    return target;
}
```

**Environment (please complete the following information):**
- Mapperly Version: 3.4.0
- .NET Version: .NET 8.0.202
- Target Framework: net8.0
- Compiler Version: 4.9.0-3.24121.1 (a98c90d5)
- C# Language Version: 12.0
- IDE: Visual Studio v17.9.3
- OS: Windows 11

### PR DESCRIPTION
**Pull request by latonz**
Closes #1182

### CODE REVIEW

File: test/Riok.Mapperly.Tests/Mapping/ObjectPropertyNullableTest.cs [hunk 1]
```diff
@@ -516,6 +518,89 @@ TestSourceBuilderOptions.Default with
...
+        var source = TestSourceBuilder.Mapping(
+            "A",
+            "B",
+            TestSourceBuilderOptions.Default with
+            {
+                ThrowOnPropertyMappingNullMismatch = true
+            },
+            "class A { public C? Value { get; set; } }",
+            "class B { public D? Value { get; set; } }",
+            "class C { public string V {get; set; } }",
```
Comment (ni507) [inline test/Riok.Mapperly.Tests/Mapping/ObjectPropertyNullableTest.cs:533]: missing space `{get; set; }`

File: test/Riok.Mapperly.Tests/Mapping/ObjectPropertyNullableTest.cs [hunk 2]
```diff
@@ -516,6 +518,89 @@ TestSourceBuilderOptions.Default with
...
+            [MapProperty("Value.Flattened", "ValueFlattened")]
+            partial B Map(A source);
+            """,
+            TestSourceBuilderOptions.Default with
+            {
+                ThrowOnPropertyMappingNullMismatch = true
+            },
+            "class A { public C? Value { get; set; } }",
+            "class B { public D? Value { get; set; } public string ValueFlattened { get; set; } }",
+            "class C { public string V {get; set; } public string Flattened { get; set; } }",
```
Comment (ni507) [inline test/Riok.Mapperly.Tests/Mapping/ObjectPropertyNullableTest.cs:575]: missing space `{get; set; }`

**Review by ni507 — CHANGES_REQUESTED** [review]

File: test/Riok.Mapperly.Tests/Mapping/ObjectPropertyNullableTest.cs (hunk 2 above)
Comment (latonz) [reply test/Riok.Mapperly.Tests/Mapping/ObjectPropertyNullableTest.cs]: fixed all of them 😊

File: test/Riok.Mapperly.Tests/Mapping/ObjectPropertyNullableTest.cs (hunk 1 above)
Comment (latonz) [reply test/Riok.Mapperly.Tests/Mapping/ObjectPropertyNullableTest.cs]: fixed all of them 😊

**Review by ni507 — APPROVED** [review]

**Pull request merged**